contracts/src/v0.8/test/HeartbeatRequester.t.sol

```diff
@@ -0,0 +1,108 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.6;
+
+import "forge-std/Test.sol";
+import {HeartbeatRequester} from "../HeartbeatRequester.sol";
+import "../mocks/MockAggregator.sol";
+import "../mocks/MockAggregatorProxy.sol";
+import "../mocks/MockOffchainAggregator.sol";
+
+// from contracts directory,
+// forge test --match-path src/v0.8/test/HeartbeatRequester.t.sol
+
+contract HeartbeatRequesterTest is Test {
+  HeartbeatRequester public heartbeatRequester;
+  MockAggregator public aggregator;
+  IAggregatorProxy public aggregatorProxy;
+  MockAggregator public aggregator2;
+  IAggregatorProxy public aggregatorProxy2;
+
+  address internal constant USER = address(2);
+
+  event HeartbeatPermitted(address indexed permittedCaller, address newProxy, address oldProxy);
+  event HeartbeatRemoved(address indexed permittedCaller, address removedProxy);
+  error HeartbeatNotPermitted();
+
+  function setUp() public {
+    heartbeatRequester = new HeartbeatRequester();
+    aggregator = new MockAggregator();
+    aggregatorProxy = IAggregatorProxy(new MockAggregatorProxy(address(aggregator)));
+    aggregator2 = new MockAggregator();
+    aggregatorProxy2 = IAggregatorProxy(new MockAggregatorProxy(address(aggregator2)));
+  }
+
+  function test_permitHeartbeat_ArbitraryUser() public {
+    vm.expectEmit(true, true, true, true);
+    emit HeartbeatPermitted(USER, address(aggregatorProxy), address(0));
+    heartbeatRequester.permitHeartbeat(USER, aggregatorProxy);
+
+    vm.expectEmit(true, true, true, true);
+    emit HeartbeatPermitted(USER, address(aggregatorProxy2), address(aggregatorProxy));
+    heartbeatRequester.permitHeartbeat(USER, aggregatorProxy2);
+  }
+
+  function test_permitHeartbeat_Deployer() public {
+    vm.expectEmit(true, true, true, true);
+    emit HeartbeatPermitted(address(this), address(aggregatorProxy), address(0));
+    heartbeatRequester.permitHeartbeat(address(this), aggregatorProxy);
+
+    vm.expectEmit(true, true, true, true);
+    emit HeartbeatPermitted(address(this), address(aggregatorProxy2), address(aggregatorProxy));
+    heartbeatRequester.permitHeartbeat(address(this), aggregatorProxy2);
+  }
+
+  function test_permitHeartbeat_NotFromOwner() public {
```

Comment:
We prefer to end all reverting tests with `Reverts` and all happy path tests with `Success` but that might be a bit opinionated. Please consider it though